### PR TITLE
CASSANDRA-16125: Allow optional query retry

### PR DESCRIPTION
**Pull request by yifan-c**
- Retry at the application level.
- Added NoRetry and ExponentialRetry strategies.
- RetryStrategy is made pluggable, as long as that the custom retry strategy extends the base and specify it in the config
- NoRetry is used if no RetryOptions is supplied in the config.

### CODE REVIEW

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                if (pausedInTotal < totalDelayMs) {
+                    return Math.min(totalDelayMs - pausedInTotal, nextMaybe); // adjust the next pause time if possible
+                }
+                return -1; // the previous retries have exhausted the permits
+            }
+        }
+
+        // Returns the total pause time according the `attempts`,i.e. [0, attempts), which is guaranteed to be greater than 0.
+        // No overflow can happen.
+        private long pausedInTotal(int attempts) {
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java:82]: this looks like a geometric sum, could probably be replaced with something like:
```
        private long pausedInTotalGeom(int attempts) {
            long result = -1 * baseDelayMs * (1 - (1L << attempts));
            return Math.min(result, totalDelayMs);
        }
```

Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java]: this looks like a geometric sum, could probably be replaced with something like:
```
        private long pausedInTotal(int attempts) {
            long result = -1 * baseDelayMs * (1 - (1L << attempts));
            return Math.min(result, totalDelayMs);
        }
```

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            } else {
+                long pausedInTotal = pausedInTotal(attempts);
+                if (pausedInTotal < totalDelayMs) {
+                    return Math.min(totalDelayMs - pausedInTotal, nextMaybe); // adjust the next pause time if possible
+                }
+                return -1; // the previous retries have exhausted the permits
+            }
+        }
+
+        // Returns the total pause time according the `attempts`,i.e. [0, attempts), which is guaranteed to be greater than 0.
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java]: greater than or equal to 0

File: common/src/main/java/org/apache/cassandra/diff/RetryStrategyFactory.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    public RetryStrategy create() {
+        if (retryOptions != null) {
+            return create(retryOptions);
+        } else {
+            logger.info("Retry is disabled.");
+            return RetryStrategy.NoRetry.INSTANCE;
+        }
+    }
+
+    public static RetryStrategy create(Map<String, String> parameters) {
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/RetryStrategyFactory.java]: feels slightly expensive doing this for every query - maybe we could make the strategy factory pluggable instead? Then just a factory.get() to get the actual strategy instance for every query?

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+public class ExponentialRetryStrategy extends RetryStrategy {
+    public final static String BASE_DELAY_MS_KEY = "base_delay_ms";
+    public final static String TOTAL_DELAY_MS_KEY = "total_delay_ms";
+    private final static String DEFAULT_BASE_DELAY_MS = String.valueOf(TimeUnit.SECONDS.toMillis(1));
+    private final static String DEFAULT_TOTAL_DELAY_MS = String.valueOf(TimeUnit.MINUTES.toMillis(30));
+
+    private final Exponential exponential;
+    private int attempts = 0;
+
+    public ExponentialRetryStrategy(Map<String, String> parameters) {
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java]: should this take `JobConfiguration.RetryOptions` instead?

File: common/src/main/java/org/apache/cassandra/diff/RetryStrategy.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class RetryStrategy {
+    private final static Logger logger = LoggerFactory.getLogger(RetryStrategy.class);
+
+    public RetryStrategy(Map<String, String> parameters) {
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/RetryStrategy.java]: `JobConfiguration.RetryOptions` ?

**Review by krummas — CHANGES_REQUESTED** [review]
looks good in general, a few comments inline

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java (hunk 1 above)
Comment (yifan-c) [reply common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategy.java]: Fancy! O(n) -> O(1). 
Yes. It is the geometric sum. Given the ratio is 2, it is actually just `2 ^ (n + 1) - 1`

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        private final static String DEFAULT_BASE_DELAY_MS = String.valueOf(TimeUnit.SECONDS.toMillis(1));
+        private final static String DEFAULT_TOTAL_DELAY_MS = String.valueOf(TimeUnit.MINUTES.toMillis(30));
+
+        private final Exponential exponential;
+        private int attempts = 0;
+
+        public ExponentialRetryStrategy(JobConfiguration.RetryOptions retryOptions) {
+            super(retryOptions);
+            long baseDelayMs = Long.parseLong(retryOptions.getOrDefault(BASE_DELAY_MS_KEY, DEFAULT_BASE_DELAY_MS));
+            long totalDelayMs = Long.parseLong(retryOptions.getOrDefault(TOTAL_DELAY_MS_KEY, DEFAULT_TOTAL_DELAY_MS));
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java]: maybe make this constructor take the 2 longs? And move the long-parsing to the `ExponentialRetryStrategyProvider` constructor

File: common/src/main/java/org/apache/cassandra/diff/RetryStrategy.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.concurrent.Callable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class RetryStrategy {
+    private final static Logger logger = LoggerFactory.getLogger(RetryStrategy.class);
+
+    public RetryStrategy(JobConfiguration.RetryOptions retryOptions) {
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/RetryStrategy.java]: can remove this constructor

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+                }
+                return -1; // the previous retries have exhausted the permits
+            }
+        }
+
+        // Returns the total pause time according to the `attempts`,
+        // i.e. [0, attempts), which is guaranteed to be greater than or equal to 0.
+        // No overflow can happen.
+        private long pausedInTotal(int attempts) {
+            if (attempts >= Long.SIZE) return totalDelayMs; // take care of overflow. Such long pause time is not realistic though.
```
Comment (krummas) [inline common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java]: ```if (attempts >= Long.numberOfLeadingZeros(baseDelayMs - 1)) return totalDelayMs;``` if we want to make sure no overflow can occur

**Review by krummas — CHANGES_REQUESTED** [review]
LGTM, just minor suggestions now, if you agree with them I can make the changes and get this committed

File: common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java (hunk 8 above)
Comment (yifan-c) [reply common/src/main/java/org/apache/cassandra/diff/ExponentialRetryStrategyProvider.java]: mostly agree. It should have taken `baseDelayMs` into consideration. 
However, I do not think the subtraction is needed. The first attempt to start to return the `total` directly is the one that could make the value wraps around, from a positive to a negative.

**Comment by krummas** [comment]
committed as https://github.com/apache/cassandra-diff/commit/4c9bc4f4e3fd7d23b1284c89266ffbf10b8f0183

**Pull request closed without merging**